src/java.base/share/classes/java/lang/foreign/MemorySegment.java

```diff
@@ -226,36 +224,41 @@
+ * full spatial, temporal and confinement bounds. To do this, clients can {@linkplain #ofAddress(long, long, MemorySession)}
+ * a native segment <em>unsafely</em>, by providing a new size, as well as a new {@linkplain MemorySession memory session}.
  * This is a <a href="package-summary.html#restricted"><em>restricted</em></a> operation and should be used with
  * caution: for instance, an incorrect segment size could result in a VM crash when attempting to dereference
  * the memory segment.
  * <p>
- * Clients requiring sophisticated, low-level control over mapped memory segments, might consider writing
+ * For example, clients requiring sophisticated, low-level control over mapped memory segments, might consider writing
  * custom mapped memory segment factories; using {@link Linker}, e.g. on Linux, it is possible to call {@code mmap}
  * with the desired parameters; the returned address can be easily wrapped into a memory segment, using
- * {@link MemoryAddress#ofLong(long)} and {@link MemorySegment#ofAddress(MemoryAddress, long, MemorySession)}.
+ * {@link #ofAddress(long, long, MemorySession)}.
  *
  * @implSpec
  * Implementations of this interface are immutable, thread-safe and <a href="{@docRoot}/java.base/java/lang/doc-files/ValueBased.html">value-based</a>.
  *
  * @since 19
  */
 @PreviewFeature(feature=PreviewFeature.Feature.FOREIGN)
-public sealed interface MemorySegment extends Addressable permits AbstractMemorySegmentImpl {
+public sealed interface MemorySegment permits AbstractMemorySegmentImpl {
 
     /**
-     * {@return the base memory address associated with this native memory segment}
-     * @throws UnsupportedOperationException if this segment is not a {@linkplain #isNative() native} segment.
-     * @throws IllegalStateException if the {@linkplain #session() session} associated with this segment is not
-     * {@linkplain MemorySession#isAlive() alive}.
-     * @throws WrongThreadException if this method is called from a thread other than the thread owning
-     * the {@linkplain #session() session} associated with this segment.
+     * Returns the base address of the memory region associated with this segment. If this memory segment is
+     * a {@linkplain #isNative() native} memory segment, then the returned address is the off-heap address
+     * at which the native memory region associated with this segment starts. If this memory segment is an array
+     * memory segment, the returned address is the byte offset into the {@linkplain #array()} object associated
+     * with this segment. In other words, the base address of an array segment is always <em>virtualized</em>.
+     *
+     * @return the base address of the memory region associated with this segment.
      */
-    @Override
-    MemoryAddress address();
+    long address();
+
+    /**
+     * {@return the Java array associated with this memory segment, if any}
+     */
+    Optional<Object> array();
```

Comment:
I wonder if array word is not too strict, I think what if we would use value object as a base for struct (sometime in future).

Reply:
> I wonder if array word is not too strict, I think what if we would use value object as a base for struct (sometime in future).

I think you are correct, the word might indeed be too strict. "Heap container" is what we're after, but I'm not sure if having a method named "heapContainer" would fly?

Reply:
I'm not good at finding names. I got used to `base`. Maybe heapBase or heapObject. I'll think about it and come back if I invite something worth of mentioning.

Reply:
I think heapContainer is good, I just thought as well about something like holder, or value holder if we would like to have something shorter - I'm really not an expert at naming.

Reply:
what about `heapBase` ?

Reply:
I'll keep the name (`array`) for now, we can always revise that later. Thanks.